File: clients/android/app/src/main/java/com/vpn/client/protocol/VpnMessage.kt
package com.vpn.client.protocol

import kotlinx.serialization.Serializable

/**
 * VPN Protocol Message Types
 *
 * Message format: [type:1byte][length:4bytes BE][payload]
 */
enum class VpnMessageType(val value: Byte) {
    AUTH_REQUEST(0x01),
    AUTH_RESPONSE(0x02),
    CONFIG_PUSH(0x03),
    KEEPALIVE(0x04),
    KEEPALIVE_ACK(0x05),
    DISCONNECT(0x06),
    DATA_PACKET(0x10);

    companion object {
        fun fromValue(value: Byte): VpnMessageType? {
            return entries.find { it.value == value }
        }
    }
}

/**
 * Authentication request payload.
 *
 * JSON format:
 * {"username":"","password":"","clientVersion":"","platform":"android"}
 */
@Serializable
data class AuthRequest(
    val username: String,
    val password: String,
    val clientVersion: String,
    val platform: String = "android"
)

/**
 * Authentication response payload.
 *
 * JSON format:
 * {"success":true,"sessionToken":"","errorMessage":""}
 */
@Serializable
data class AuthResponse(
    val success: Boolean,
    val sessionToken: String = "",
    val errorMessage: String? = null
)

/**
 * Configuration request (sent after successful auth).
 */
@Serializable
data class ConfigRequest(
    val sessionToken: String
)

/**
 * VPN configuration pushed by server.
 *
 * JSON format:
 * {"assignedIP":"","subnetMask":"","gateway":"","dns":[],"mtu":1400}
 */
@Serializable
data class VpnConfig(
    val assignedIP: String,
    val subnetMask: String,
    val gateway: String,
    val dns: List<String>,
    val mtu: Int = 1400
)

/**
 * Keepalive message (empty payload, just the type).
 */
object Keepalive

/**
 * Disconnect message (empty payload, just the type).
 */
object Disconnect

/**
 * Data packet container.
 * The payload is raw IP packet data.
 */
data class DataPacket(
    val data: ByteArray
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as DataPacket
        return data.contentEquals(other.data)
    }

    override fun hashCode(): Int {
        return data.contentHashCode()
    }
}

/**
 * Generic VPN message wrapper.
 */
sealed class VpnMessage {
    data class Auth(val request: AuthRequest) : VpnMessage()
    data class AuthResp(val response: AuthResponse) : VpnMessage()
    data class Config(val config: VpnConfig) : VpnMessage()
    data object KeepAlive : VpnMessage()
    data object KeepAliveAck : VpnMessage()
    data object DisconnectMsg : VpnMessage()
    data class Data(val packet: DataPacket) : VpnMessage()
}

/**
 * Connection statistics.
 */
data class ConnectionStats(
    val bytesReceived: Long = 0,
    val bytesSent: Long = 0,
    val packetsReceived: Long = 0,
    val packetsSent: Long = 0,
    val connectedAt: Long = 0,
    val lastActivityAt: Long = 0
)

/**
 * VPN connection state for UI.
 */
data class VpnState(
    val isConnected: Boolean = false,
    val isConnecting: Boolean = false,
    val assignedIp: String = "",
    val serverAddress: String = "",
    val stats: ConnectionStats = ConnectionStats(),
    val error: String? = null
)

File: clients/windows/VPNClient/Protocol/VpnMessage.cs
namespace VPNClient.Protocol;

/// <summary>
/// VPN protocol message types
/// </summary>
public enum MessageType : byte
{
    /// <summary>
    /// Authentication request from client
    /// Payload: JSON {"username":"","password":"","clientVersion":"","platform":"windows"}
    /// </summary>
    AuthRequest = 0x01,

    /// <summary>
    /// Authentication response from server
    /// Payload: JSON {"success":true,"sessionToken":"","errorMessage":""}
    /// </summary>
    AuthResponse = 0x02,

    /// <summary>
    /// Configuration push from server after successful authentication
    /// Payload: JSON {"assignedIP":"","subnetMask":"","gateway":"","dns":[],"mtu":1400}
    /// </summary>
    ConfigPush = 0x03,

    /// <summary>
    /// Keepalive message sent periodically
    /// Payload: empty
    /// </summary>
    Keepalive = 0x04,

    /// <summary>
    /// Keepalive acknowledgment
    /// Payload: empty
    /// </summary>
    KeepaliveAck = 0x05,

    /// <summary>
    /// Disconnect request
    /// Payload: empty
    /// </summary>
    Disconnect = 0x06,

    /// <summary>
    /// Data packet containing raw IP packet
    /// Payload: raw IP packet bytes
    /// </summary>
    DataPacket = 0x10
}

/// <summary>
/// VPN protocol message structure
/// Header format: [type:1byte][length:4bytes BE][payload]
/// </summary>
public class VpnMessage
{
    /// <summary>
    /// Message type identifier
    /// </summary>
    public MessageType Type { get; set; }

    /// <summary>
    /// Message payload (raw bytes)
    /// </summary>
    public byte[] Payload { get; set; } = Array.Empty<byte>();

    /// <summary>
    /// Total message length including header
    /// </summary>
    public int TotalLength => 5 + Payload.Length; // 1 byte type + 4 bytes length + payload
}

/// <summary>
/// Authentication request payload structure
/// </summary>
public class AuthRequest
{
    /// <summary>
    /// Username for authentication
    /// </summary>
    public string Username { get; set; } = string.Empty;

    /// <summary>
    /// Password for authentication
    /// </summary>
    public string Password { get; set; } = string.Empty;

    /// <summary>
    /// Client version string
    /// </summary>
    public string ClientVersion { get; set; } = "1.0.0";

    /// <summary>
    /// Platform identifier (always "windows" for this client)
    /// </summary>
    public string Platform { get; set; } = "windows";
}

/// <summary>
/// Authentication response payload structure
/// </summary>
public class AuthResponse
{
    /// <summary>
    /// Whether authentication was successful
    /// </summary>
    public bool Success { get; set; }

    /// <summary>
    /// Session token for authenticated session
    /// </summary>
    public string? SessionToken { get; set; }

    /// <summary>
    /// Error message if authentication failed
    /// </summary>
    public string? ErrorMessage { get; set; }
}

/// <summary>
/// Configuration push payload structure
/// </summary>
public class ConfigPush
{
    /// <summary>
    /// Assigned IP address for the VPN interface
    /// </summary>
    public string? AssignedIP { get; set; }

    /// <summary>
    /// Subnet mask for the VPN interface
    /// </summary>
    public string? SubnetMask { get; set; }

    /// <summary>
    /// Gateway IP address
    /// </summary>
    public string? Gateway { get; set; }

    /// <summary>
    /// DNS server addresses
    /// </summary>
    public string[]? Dns { get; set; }

    /// <summary>
    /// Maximum transmission unit size
    /// </summary>
    public int Mtu { get; set; } = 1400;
}
